File: Rapptr_iOS_Test/iOS Programmer Test/iOSTest/LoginClient.m
//
//  LoginClient.m
//  iOSTest
//
//  Created by Sajan Shrestha on 2/12/21.
//  Copyright © 2021 D&ATechnologies. All rights reserved.
//

#import "LoginClient.h"

@interface LoginClient ()
@property (nonatomic, strong) NSURLSession *session;
@end

@implementation LoginClient

/**
 * =========================================================================================
 * INSTRUCTIONS
 * =========================================================================================
 * 1) Make a request here to login.
 *
 * 2) Using the following endpoint, make a request to login
 *    URL: http://dev.rapptrlabs.com/Tests/scripts/login.php
 *
 * 3) Don't forget, the endpoint takes two parameters 'email' and 'password'
 *
 * 4) email - info@rapptrlabs.com
 *   password - Test123
 **/

- (void)loginWithEmail:(NSString *)email password:(NSString *)password completion:(void (^)(NSDictionary *))completion
{
    
    NSDate *start = [NSDate date];
    
    NSString *urlString = @"http://dev.rapptrlabs.com/Tests/scripts/login.php";
    NSURLComponents *components = [NSURLComponents componentsWithString:urlString];
    
    NSURLQueryItem *emailQueryItem = [NSURLQueryItem queryItemWithName:@"email" value:email];
    NSURLQueryItem *passwordQueryItem = [NSURLQueryItem queryItemWithName:@"password" value:password];
    NSArray* queryItems = @[ emailQueryItem, passwordQueryItem ];
    
    components.queryItems = queryItems;
    
    NSURL *url = [components URL];
    
    [[NSURLSession.sharedSession dataTaskWithURL:url completionHandler:^(NSData * _Nullable data, NSURLResponse * _Nullable response, NSError * _Nullable error) {
                
        if (response) {
            NSTimeInterval timeInterval = [start timeIntervalSinceNow];
            double timeIntervalInMs = timeInterval * 1000;
            NSNumber *networkingTime = [NSNumber numberWithDouble:timeIntervalInMs];
            
            completion(@{@"Success" : @YES, @"NetworkingTime": networkingTime});
        }
        else {
            completion(@{@"Success" : @NO});
        }
            
    }] resume];

}

@end

File: Rapptr_iOS_Test/iOS Programmer Test/iOSTest/LoginClient.swift
//
//  LoginClient.swift
//  iOSTest
//
//  Created by Sajan Shrestha on 2/10/21.
//  Copyright © 2021 D&ATechnologies. All rights reserved.
//

import Foundation

struct LoginClient {
    
    
        
    static func login(email: String, password: String, completion: @escaping (Bool, TimeInterval?) -> Void) {
        
        let startDate = Date()
        
        guard var urlComponents = URLComponents(string: "http://dev.rapptrlabs.com/Tests/scripts/login.php") else { return }
        let emailQuery = URLQueryItem(name: "email", value: email)
        let passwordQuery = URLQueryItem(name: "password", value: password)
        urlComponents.queryItems = [emailQuery, passwordQuery]
                
        guard let url = urlComponents.url else { return }
        print(url)
        let request = URLRequest(url: url)
        URLSession.shared.dataTask(with: request) { (data, response, error) in
            if let _ = response {
                let executionTime = Date().timeIntervalSince(startDate)
                completion(true, executionTime)
            }
            else {
                completion(false, nil)
            }
        }.resume()
    }
}
